src/data/list/of_fn.lean

```diff
@@ -157,4 +159,25 @@ by simp only [mem_of_fn, set.forall_range_iff]
   of_fn (λ i : fin n, c) = repeat c n :=
 nat.rec_on n (by simp) $ λ n ihn, by simp [ihn]
 
+/-- Lists are equivalent to the sigma type of tuples of a given length. -/
+@[simps]
+def equiv_sigma_tuple : list α ≃ Σ n, fin n → α :=
+{ to_fun := λ l, ⟨l.length, λ i, l.nth_le ↑i i.2⟩,
+  inv_fun := λ f, list.of_fn f.2,
+  left_inv := list.of_fn_nth_le,
+  right_inv := λ ⟨n, f⟩, fin.sigma_eq_of_eq_comp_cast (length_of_fn _) $ funext $ λ i,
+    nth_le_of_fn' f i.prop }
+
+/-- A recursor for lists that expands a list into a function mapping to its elements. -/
+def of_fn_rec (C : list α → Sort*) (h : Π n (f : fin n → α), C (list.of_fn f)) (l : list α) : C l :=
```

Comment:
Should it be `@[elab_as_eliminator]`? Does it work with the `induction` tactic? And there should be a lemma saying that `of_fn_rec_apply : of_fn_rec C h (of_fn f) = C f`. Otherwise LGTM. Thanks!
bors d+

Reply:
My proof for `of_fn_rec_apply` seems awful; do you have any suggestions for making it less so?

Reply:
The following generic lemma might help.
```
import logic.equiv.basic

universes u v

lemma equiv.rec_on_symm_apply_apply {α β : Sort u} {γ : α → Sort v} (e : α ≃ β)
  (f : Π b : β, γ (e.symm b)) (y : β) :
  (λ x, eq.rec_on (e.symm_apply_apply x) (f (e x)) : Π x : α, γ x) (e.symm y) = f y :=
eq_of_heq $ (eq_rec_heq _ _).trans $ by rw [e.apply_symm_apply]
```

Reply:
Yes, that lemma did the trick; thanks! I've pushed four variants of that lemma.